django_cloud_deploy/cli/prompt_manager.py

```diff
@@ -0,0 +1,673 @@
+
+
+class TemplatePromptManager(object):
+    """Used as a base template for all Prompt Managers interacting with user.
+
+    They should all have prompt method that should make subsequent prompt calls.
+    Manager should own only one parameter, for which they should have a
+    validate function for which they will call if the argument is already
+    supplied. They also handle any busines logic related to a prompt, as well
+    as become instantiated with any API related classes.
+
+    """
+
+    # Parameter Manager Owns must be set
+    PARAMETER = None
+
+    def __init__(self, console: io.IO, step: str, args):
+        self.console = console
+        self.step = step
+        self.args = args
+
+    def prompt(self) -> None:
+        pass
+
+    def validate(self, val: Any) -> bool:
+        return True
+
+    def get_credentials(self):
+        return self.args.get('credentials', None)
+
+    def _is_valid_passed_arg(self) -> bool:
+        value = self.args.get(self.PARAMETER, None)
+        if value is None:
+            return False
+
+        try:
+            if self.validate(value):
+                msg = '{} {}: {}'.format(self.step, self.PARAMETER, value)
+                self.console.tell(msg)
+                return True
+        except ValueError as e:
+            self.console.error(e)
+            return False
+
+        return False
+
+
+class StringTemplatePromptManager(TemplatePromptManager):
+    """Template for a simple string Prompt Manager."""
+
+    PARAMETER = ''
+    PARAMTER_PRETTY = ''
+    DEFAULT_VALUE = ''
+    BASE_MESSAGE = '{} Enter a value for {} or leave blank to use'
+    DEFAUlT_MESSAGE = '[{}]: '
+
+    def _generate_name_prompt(self):
+        base_message = self.BASE_MESSAGE.format(self.step, self.PARAMTER_PRETTY)
+        default_message = self.DEFAUlT_MESSAGE.format(self.DEFAULT_VALUE)
+        msg = '\n'.join([base_message, default_message])
+        return prompt_new.AskPrompt(
+            msg, self.console, self.validate, default=self.DEFAULT_VALUE)
+
+    def prompt(self):
+        if self._is_valid_passed_arg():
+            return
+
+        self.args[self.PARAMETER] = self._generate_name_prompt().prompt()
+
+
+class GoogleProjectNameManager(TemplatePromptManager):
+
+    PARAMETER = 'project_name'
+
+    def __init__(self,
+                 console: io.IO,
+                 step: str,
+                 args,
+                 project_client: Optional[project.ProjectClient] = None):
+        super().__init__(console, step, args)
+        creds = self.get_credentials()
+        self.project_client = (project_client or
+                               project.ProjectClient.from_credentials(creds))
+
+    def validate(self, s: str) -> bool:
+        """Validates that a string is a valid project name.
+
+        Args:
+            s: The string to validate.
+
+        Raises:
+            ValueError: if the input string is not valid.
+        """
+        if not (4 <= len(s) <= 30):
```

Comment:
I think that there is more to it than that.

Reply:
I believe I copied our previous code. But dont mind adding extra logic.